### PR TITLE
application: serial_lte_modem: Add util function for DNS lookup

### PR DESCRIPTION
**Pull request by junqingzou**
Add util_resolve_host() to resolve remote host by name or address.
Prepared for future multiple PDN support.

Signed-off-by: Jun Qing Zou <jun.qing.zou@nordicsemi.no>

### CODE REVIEW

File: applications/serial_lte_modem/src/slm_util.c [hunk 1]
```diff
@@ -221,6 +222,49 @@ int util_str_to_int(const char *str_buf, int base, int *output)
...
+	struct addrinfo *ai = NULL;
+	struct addrinfo hints = {
+		.ai_family = family
+	};
+
+	if (sa == NULL) {
+		return DNS_EAI_AGAIN;
+	}
+
+	if (family != AF_UNSPEC) {
```
Comment (rlubos) [inline applications/serial_lte_modem/src/slm_util.c:246]: Hmm, isn't it just possible to provide hints in any case with family set to `AF_UNSPEC`? It should return either IPv4 or 6 in this case. Or is there some bug in nrf_modem I forgot about?

**Review by rlubos — COMMENTED** [review]
Looks good, just some suggestions.

File: applications/serial_lte_modem/src/slm_util.c [hunk 2]
```diff
@@ -221,6 +222,49 @@ int util_str_to_int(const char *str_buf, int base, int *output)
 	}
 
 	*output = temp;
+	return 0;
+}
+
+/**
+ * @brief Resolve remote host by hostname or IP address
+ */
+#define SIN6(A) ((struct sockaddr_in6 *)(A))
```
[the remark this answers is not in the document]
Comment (junqingzou) [reply applications/serial_lte_modem/src/slm_util.c:231]: This code is not required after adding the service string

File: applications/serial_lte_modem/src/slm_util.c [hunk 3]
```diff
@@ -221,6 +222,49 @@ int util_str_to_int(const char *str_buf, int base, int *output)
...
+	if (err) {
+		return err;
+	}
+
+	*sa = *(ai->ai_addr);
+	freeaddrinfo(ai);
+
+	switch (sa->sa_family) {
+	case AF_INET6:
+		SIN6(sa)->sin6_port = htons(port);
```
[the remark this answers is not in the document]
Comment (junqingzou) [reply applications/serial_lte_modem/src/slm_util.c:260]: Great suggestion, tested to be good for both IPv4 and IPv6, service port auto-filled

File: applications/serial_lte_modem/src/slm_util.c (hunk 1 above)
Comment (junqingzou) [reply applications/serial_lte_modem/src/slm_util.c:246]: It's not an issue with nrf_modem any more (after fix my JIRA case).
Anyway now `hints` is always required for `ai_flags`

**Review by rlubos — APPROVED** [review]
LGTM

**Review by lats1980 — COMMENTED** [review]
Does the AT#XGETADDRINFO also support multiple PDN in the future? It does not use new DNS lookup utility.

They are different use cases:
.AT#XGETADDRINFO use default primary PDN to look up given hostname and list all resolved result.
.util_resolve_host() use desired address family to look up an host (on specified PDN in the future) expecting only one resolved result, for example use IPv4 or IPv6 socket on secondary PDP.

**Comment by junqingzou** [comment]
@rlubos @lats1980 I added the AI flag of PDN selection in util_resolve_host().
Please review again. NOTE this is Nordic/NCS-only feature.

**Review by FrancescoSer — APPROVED** [review]

**Review by lats1980 — APPROVED** [review]
LGTM

**Pull request merged**